File: go/0015.3sum/solution.go
// Created by Noirkelo at 2025/08/05 14:18
// leetgo: dev
// https://leetcode.cn/problems/3sum/

package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func threeSum(nums []int) (ans [][]int) {
	//a_i + a_j =-a_k
	//i  | j | k
	//i  | 忽略 | j | k
	//
	//假设i j 固定 找 a_k
	// 1 1 -2
	len_nums := len(nums)
	sort.Ints(nums)
	for i := 0; i < len_nums; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		for j, k := i+1, len_nums-1; j < len_nums; j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}
			//nums[j]+nums[k]只会变得更小

			for j < k && nums[j]+nums[k]+nums[i] > 0 {
				k--
			}
			if j == k {
				break
			}
			if nums[j]+nums[k]+nums[i] == 0 {
				ans = append(ans, []int{nums[i], nums[j], nums[k]})
			}
		}
	}
	return
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	ans := threeSum(nums)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0015.3sum/solution.rs
// Created by Noirkelo at 2025/08/05 14:18
// leetgo: dev
// https://leetcode.cn/problems/3sum/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin

impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
		let nums_len=nums.len();
		let mut  ans= Vec::<Vec<i32>>::default();
		nums.sort();
		for i in 0..nums_len {
			if i>0 && nums[i-1]==nums[i]{
				continue
			}
			let mut k=nums.len()-1;
			for j in i+1..nums_len{
				if j>i+1 && nums[j-1]==nums[j]{
					continue
				}
				while j<k&& nums[j]+nums[k]+nums[i]>0{
					k-=1;
				}
				//j!=k 本身也不能当成答案
				if j==k{
					break;
				}
				//提取其中一个 ，利用有序性定下来k和j的遍历顺序
				if nums[j]+nums[k]+nums[i]==0{
					ans.push(vec![nums[i],nums[j],nums[k]]);
				}


			}
		}
		ans
    }
}

// @lc code=end

fn main() -> Result<()> {
	let nums: Vec<i32> = deserialize(&read_line()?)?;
	let ans: Vec<Vec<i32>> = Solution::three_sum(nums).into();

	println!("\noutput: {}", serialize(ans)?);
	Ok(())
}
